File: api/app/models/article.py
from sqlalchemy_utils import UUIDType
import uuid
from sqlalchemy import Column, String, Text, DateTime, ForeignKey
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
from app import db
from sqlalchemy.dialects.postgresql import ARRAY
from sqlalchemy.ext.mutable import MutableList


class Article(db.Model):
    __tablename__ = "articles"

    id = Column(UUIDType(binary=False), primary_key=True, default=uuid.uuid4)
    title = Column(String, nullable=False)
    snippet = Column(Text, nullable=False)
    summary = Column(Text, nullable=True)
    # body = Column(Text, nullable=True)
    body_original = Column(Text, nullable=True)
    body_translated = Column(Text, nullable=True)
    source_url = Column(Text, nullable=True)
    published_date = Column(DateTime(timezone=True), nullable=True)
    cover_image = Column(Text, nullable=True)

    countries = Column(MutableList.as_mutable(ARRAY(String)), nullable=False)
    language = Column(String(255), nullable=False)

    tenant_id = Column(UUIDType(binary=False), ForeignKey("tenants.id"), nullable=False)
    tenant = relationship("Tenant", backref="articles")

    created_at = Column(DateTime(timezone=True), default=func.now(), nullable=False)
    updated_at = Column(
        DateTime(timezone=True),
        default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )

    def __repr__(self):
        return f"<Article {self.title}>"

File: frontend/src/graphql/queries/article.ts
import { gql } from "@apollo/client";

const GENERATE_ARTICLES = gql`
  mutation GenerateArticles($input: GenerateArticlesInput!) {
    generateArticles(input: $input) {
      articles {
        id
      }
    }
  }
`;

const GET_ARTICLE = gql`
  query GetArticle($id: ID!) {
    article(id: $id) {
      id
      title
      snippet
      summary
      bodyOriginal
      bodyTranslated
      sourceUrl
      publishedDate
      coverImage
      countries
      language
      createdAt
    }
  }
`;

const LIST_ARTICLES = gql`
  query ListArticles {
    articles {
      id
      title
      snippet
      summary
      bodyOriginal
      bodyTranslated
      sourceUrl
      publishedDate
      coverImage
      countries
      language
      createdAt
    }
  }
`;

const ADD_SUMMARY_TO_ARTICLE = gql`
  mutation AddSummaryToArticle($input: AddSummaryToArticleInput!) {
    addSummaryToArticle(input: $input) {
      article {
        id
        summary
      }
    }
  }
`;

const ADD_TRANSLATION_TO_ARTICLE = gql`
  mutation AddTranslationToArticle($input: AddTranslationToArticleInput!) {
    addTranslationToArticle(input: $input) {
      article {
        id
        bodyTranslated
      }
    }
  }
`;

const DELETE_ARTICLE = gql`
  mutation DeleteArticle($id: ID!) {
    deleteArticle(id: $id) {
      success
    }
  }
`;

export {
  GENERATE_ARTICLES,
  LIST_ARTICLES,
  ADD_SUMMARY_TO_ARTICLE,
  ADD_TRANSLATION_TO_ARTICLE,
  GET_ARTICLE,
  DELETE_ARTICLE,
};
